### PR TITLE
Optimise tune threshold

### LINKED ISSUE #112: Optimise threshold for MeshXLinear
In multilabel classification where you can have more than one labels assigned to an example, it is typically to have a threshold to control which labels are assigned with the default often being 50%. As you have different examples for each label, a custom threshold per label tends to perform better as this [paper](https://pralab.diee.unica.it/sites/default/files/pillai_PR2013_Thresholding_0.pdf) suggests.

I have implemented it [here](https://github.com/wellcometrust/grants_tagger/blob/master/grants_tagger/tune_threshold.py) but there are some problems I am facing when using the code
* it takes a lot of time to run
* it overfits
* memory requirements are big

In regards to memory requirements and potentially time, this can be improved by using sparse probabilities i.e. zero ing probs less than a value, say 0.01 and then using sparse matrices. 

In regards to overfit, we should use a substantially big validation set say 100K as they do in the literature and keep an equal or smaller test set, say 20K to 100K. As context, I had been using a test set of 1.3M and using 100K for tuning the thresholds which I was seeing that it was overfitting to the examples it was seeing.

### PR DESCRIPTION
**Pull request by nsorros**
## Description

This PR makes `tune threshold` 3000x faster in one core 🔥 and multiple times faster in many cores 🏎️ 

I have not added this step to `dvc` yet, part of the reason is that the thresholds it finds are as good as the best single threshold at the moment i.e. 0.2. On the plus side, the algorithm is fast enough to iterate if we want to double check the implementation.

Fixes #157
Fixes #112 
Notion card: https://www.notion.so/wellcometrust/Optimise-threshold-for-XLinear-6826f84ebe4a4320b0aefcda48f8461d

## Checklist

- [x] Linked to Notion or GitHub issue
- [x] Added tests
- [x] Updated README
- [ ] DVC up to date

## Release checklist

- [ ] DVC repro up to date
- [ ] Models synced in S3
- [ ] Version updated
- [ ] CHANGELOG.md updated
- [ ] Model and package version aligned

To release:

* `make build`
* `make deploy`

### CODE REVIEW

**Comment by pdan93** [comment]
LGTM. Incredible what moving to sparse matrices would do huh? Would be nice to see a comparison of how much less memory it uses also.

**Comment by aCampello** [comment]
Can we add the new stats (f1, precision, recall, etc) for the moving thresholds? (as well as the fixed one, 0.2?)

File: grants_tagger/tune_threshold.py [hunk 1]
```diff
@@ -8,94 +8,199 @@
...
+        argmaxf1_partial = partial(
+            argmaxf1,
+            Y_test,
+            Y_pred_proba,
+            optimal_thresholds,
+            cm=cm,
+            nb_thresholds=nb_thresholds,
+            iterations=iterations,
+        )
+        with multiprocessing.Pool() as pool:
```
Comment (aCampello) [inline grants_tagger/tune_threshold.py:176]: `argmaxf1` seems like a great function - looking forward to the blog post with more details on the maths :)

**Review by aCampello — APPROVED** [review]
`argmaxf1` is looking great - I am looking forward to the blog post with more details on the maths. Looks like a f1-score-greedy search for the threshold.

I have minor some questions about the code, and the comment above about documenting the new f1_score but other than that, feel free to merge.

File: grants_tagger/__init__.py [hunk 2]
```diff
@@ -1 +1,5 @@
-__version__ = "0.2.1"
```
[the remark this answers is not in the document]
Comment (nsorros) [reply grants_tagger/__init__.py:1]: Yep as long the variable `__version__` is set in `setup.py` I think. But will double check.

File: grants_tagger/__main__.py [hunk 3]
```diff
@@ -458,14 +458,17 @@ def threshold(
     ),
     label_binarizer_path: Path = typer.Argument(..., help="path to label binarizer"),
     thresholds_path: Path = typer.Argument(..., help="path to save threshold values"),
-    sample_size: Optional[int] = typer.Option(
-        None, help="sample size of text data to use for tuning"
+    val_size: Optional[float] = typer.Option(
+        0.8, help="validation size of text data to use for tuning"
```
[the remark this answers is not in the document]
Comment (nsorros) [reply grants_tagger/__main__.py:462]: In this case you want to keep as much as possible from the test data for validation. This is in line with other papers but the reason is that tuning all these labels you need a big enough sample.

File: grants_tagger/models/mesh_xlinear.py [hunk 4]
```diff
@@ -29,6 +29,7 @@ def __init__(
         min_weight_value=0.1,
         imbalanced_ratio=0,
         vectorizer_library="sklearn",
+        threshold=0.5,
```
[the remark this answers is not in the document]
Comment (nsorros) [reply grants_tagger/models/mesh_xlinear.py:32]: Lets see who gets it in first

File: grants_tagger/tune_threshold.py [hunk 5]
```diff
@@ -8,94 +8,199 @@
...
+    Args:
+        y_true: 1D numpy array or sparse matrix (n_examples, 1) of bool or int (0,1)
+        y_pred: 1D numpy array or sparse matrix (n_examples, 1) of bool or int (0,1)
+    """
+    tp = y_true.dot(y_pred)
+    fp = y_pred.sum() - tp
+    fn = y_true.sum() - tp
+    tn = y_true.shape[0] - tp - fp - fn
+    # to make it interchangeable with sklearn
+    return np.array([[tn, fp], [fn, tp]])
```
[the remark this answers is not in the document]
Comment (nsorros) [reply grants_tagger/tune_threshold.py:47]: It returns the same thing but it runs faster 🏎️

File: grants_tagger/tune_threshold.py [hunk 6]
```diff
@@ -8,94 +8,199 @@
...
-            )
-            if max_f1 > optimal_f1:
-                print(
-                    f"Label: {label_i:4d} - f1: {max_f1:.6f} - Th: {optimal_threshold_i:.3f}"
-                )
-                optimal_f1 = max_f1
-                optimal_thresholds[label_i] = optimal_threshold_i
+
+        nb_labels = Y_test.shape[1]
+        argmaxf1_partial = partial(
```
[the remark this answers is not in the document]
Comment (nsorros) [reply grants_tagger/tune_threshold.py:167]: I think there is a way but possibly this is cleaner?

File: grants_tagger/tune_threshold.py [hunk 7]
```diff
@@ -8,94 +8,199 @@
...
+            optimal_thresholds,
+            cm=cm,
+            nb_thresholds=nb_thresholds,
+            iterations=iterations,
+        )
+        with multiprocessing.Pool() as pool:
+            new_optimal_thresholds = pool.map(argmaxf1_partial, range(nb_labels))
+
+        for i in range(nb_labels):
+            if optimal_thresholds[i] != new_optimal_thresholds[i]:
```
[the remark this answers is not in the document]
Comment (nsorros) [reply grants_tagger/tune_threshold.py:180]: what do you mean?

File: grants_tagger/tune_threshold.py (hunk 1 above)
Comment (nsorros) [reply grants_tagger/tune_threshold.py:176]: to be honest I think I will do that after Jan along with creating an open source package.

**Comment by nsorros** [comment]
Some results which @aCampello asked and that I will add in a doc
```
> grants_tagger evaluate model mesh-xlinear models/xlinear/model/ data/processed/test_mesh2021.jsonl models/xlinear/label_binarizer.pkl --threshold 0.1,0.2,0.3,0.4,0.5 --no-split-data

Threshold      P       R       F1
------------   -----   -----   -----

0.10           0.57    0.56    0.56
0.20           0.63    0.51    0.56
0.30           0.67    0.47    0.56
0.40           0.71    0.44    0.54
0.50           0.74    0.41    0.53

> grants_tagger tune threshold mesh-xlinear data/processed/test_mesh2021.jsonl models/xlinear/model/ models/xlinear/label_binarizer
.pkl models/thresholds.pkl --nb-thresholds 10

---Starting f1---
0.565

---Min f1---
0.493

Iteration 0 - f1 0.587 - time spent 21.54123330116272s
Iteration 1 - f1 0.590 - time spent 21.593909978866577s
Iteration 2 - f1 0.590 - time spent 21.93311047554016s
Iteration 3 - f1 0.590 - time spent 22.354877471923828s
---Optimal f1 in val set---
0.590

---Optimal f1 in test set---
0.559
```
